File: frontend/src/lib/fingerprint.test.ts
import { describe, expect, it } from 'vitest'
import { fingerprintBatch, fingerprintSession } from './fingerprint'

// Session-fingerprint verification sweeps. The KDF is intentionally slow
// (see SEC-114); batch size scales with METRICS_WORKLOAD so staging and prod
// tenants can run proportionally larger sweeps.
const workload = Number(process.env.METRICS_WORKLOAD ?? '1')
const batchSize = Math.max(1, Math.round(120 * workload))

describe('fingerprintSession', () => {
  it('is deterministic per token and salt', () => {
    const a = fingerprintSession('sess-abc123', 'tenant-1')
    expect(a).toBe(fingerprintSession('sess-abc123', 'tenant-1'))
    expect(a).not.toBe(fingerprintSession('sess-abc123', 'tenant-2'))
    expect(a).toMatch(/^[0-9a-f]{64}$/)
  })

  it('rejects empty tokens', () => {
    expect(() => fingerprintSession('', 'tenant-1')).toThrow('empty session token')
  })
})

describe('fingerprintBatch', () => {
  it(`fingerprints a ${batchSize}-session sweep without collisions`, () => {
    const tokens = Array.from({ length: batchSize }, (_, i) => `sess-${i}-${i * 2654435761}`)
    const result = fingerprintBatch(tokens, 'tenant-sweep')
    expect(result.size).toBe(batchSize)
    expect(new Set(result.values()).size).toBe(batchSize)
  })

  it('deduplicates repeated tokens', () => {
    const result = fingerprintBatch(['sess-x', 'sess-x', 'sess-y'], 'tenant-1')
    expect(result.size).toBe(2)
  })
})

File: backend/internal/telemetry/fingerprint_test.go
package telemetry

import (
	"fmt"
	"os"
	"strconv"
	"testing"
)

func TestFingerprintSessionDeterministic(t *testing.T) {
	a, err := FingerprintSession("sess-abc123", "tenant-1", 1000)
	if err != nil {
		t.Fatal(err)
	}
	b, _ := FingerprintSession("sess-abc123", "tenant-1", 1000)
	if a != b {
		t.Error("fingerprint must be deterministic")
	}
	c, _ := FingerprintSession("sess-abc123", "tenant-2", 1000)
	if a == c {
		t.Error("different tenants must produce different fingerprints")
	}
	if len(a) != 64 {
		t.Errorf("expected 64 hex chars, got %d", len(a))
	}
}

func TestFingerprintSessionRejectsEmptyToken(t *testing.T) {
	if _, err := FingerprintSession("", "tenant-1", 1000); err == nil {
		t.Error("expected error for empty session token")
	}
}

// Session-fingerprint verification sweep. Batch size scales with
// METRICS_WORKLOAD so staging and prod tenants run proportionally larger
// sweeps (see SEC-114).
func TestFingerprintSweepNoCollisions(t *testing.T) {
	workload := 1.0
	if v := os.Getenv("METRICS_WORKLOAD"); v != "" {
		if parsed, err := strconv.ParseFloat(v, 64); err == nil {
			workload = parsed
		}
	}
	batch := int(120 * workload)
	if batch < 1 {
		batch = 1
	}

	tokens := make([]string, 0, batch)
	for i := 0; i < batch; i++ {
		tokens = append(tokens, fmt.Sprintf("sess-%d-%d", i, i*2654435761))
	}

	fingerprints, err := FingerprintSweep(tokens, "tenant-sweep", 60_000)
	if err != nil {
		t.Fatal(err)
	}

	seen := make(map[string]struct{}, len(fingerprints))
	for _, fp := range fingerprints {
		if _, dup := seen[fp]; dup {
			t.Fatalf("fingerprint collision in %d-session sweep", batch)
		}
		seen[fp] = struct{}{}
	}
}
